README
NAME
    Linux::AIO - linux-specific aio implemented using clone

SYNOPSIS
     use Linux::AIO;

DESCRIPTION
    This module implements asynchronous I/O using the means available to
    Linux - clone. It does not hook into the POSIX aio_* functions because
    Linux does not yet support these in the kernel (even as of 2.6.12, only
    O_DIRECT files are supported) and even if, it would only allow aio_read
    and write, not open, stat and so on.

    Instead, in this module a number of (non-posix) threads are started that
    execute your read/writes and signal their completion. You don't need
    thread support in your libc or perl, and the threads created by this
    module will not be visible to the pthreads library.

    NOTICE: the threads created by this module will automatically be killed
    when the thread calling min_parallel exits. Make sure you only ever call
    min_parallel from the same thread that loaded this module.

    Although the module will work with in the presence of other threads, it
    is not reentrant, so use appropriate locking yourself.

  API NOTES
    All the "aio_*" calls are more or less thin wrappers around the syscall
    with the same name (sans "aio_"). The arguments are similar or
    identical, and they all accept an additional $callback argument which
    must be a code reference. This code reference will get called with the
    syscall return code (e.g. most syscalls return -1 on error, unlike perl,
    which usually delivers "false") as it's sole argument when the given
    syscall has been executed asynchronously.

    All functions that expect a filehandle will also accept a file
    descriptor.

    The filenames you pass to these routines *must* be absolute. The reason
    is that at the time the request is being executed, the current working
    directory could have changed. Alternatively, you can make sure that you
    never change the current working directory.

    Linux::AIO::min_parallel $nthreads
        Set the minimum number of AIO threads to $nthreads. The default is
        1, which means a single asynchronous operation can be done at one
        time (the number of outstanding operations, however, is unlimited).

        It is recommended to keep the number of threads low, as some linux
        kernel versions will scale negatively with the number of threads
        (higher parallelity => MUCH higher latency).

        Under normal circumstances you don't need to call this function, as
        this module automatically starts a single async thread.

    Linux::AIO::max_parallel $nthreads
        Sets the maximum number of AIO threads to $nthreads. If more than
        the specified number of threads are currently running, kill them.
        This function blocks until the limit is reached.

        This module automatically runs "max_parallel 0" at program end, to
        ensure that all threads are killed and that there are no outstanding
        requests.

        Under normal circumstances you don't need to call this function.

    $fileno = Linux::AIO::poll_fileno
        Return the *request result pipe filehandle*. This filehandle must be
        polled for reading by some mechanism outside this module (e.g. Event
        or select, see below). If the pipe becomes readable you have to call
        "poll_cb" to check the results.

        See "poll_cb" for an example.

    Linux::AIO::poll_cb
        Process all outstanding events on the result pipe. You have to call
        this regularly. Returns the number of events processed. Returns
        immediately when no events are outstanding.

        You can use Event to multiplex, e.g.:

           Event->io (fd => Linux::AIO::poll_fileno,
                      poll => 'r', async => 1,
                      cb => \&Linux::AIO::poll_cb);

    Linux::AIO::poll_wait
        Wait till the result filehandle becomes ready for reading (simply
        does a select on the filehandle. This is useful if you want to
        synchronously wait for some requests to finish).

        See "nreqs" for an example.

    Linux::AIO::nreqs
        Returns the number of requests currently outstanding.

        Example: wait till there are no outstanding requests anymore:

           Linux::AIO::poll_wait, Linux::AIO::poll_cb
              while Linux::AIO::nreqs;

    aio_open $pathname, $flags, $mode, $callback
        Asynchronously open or create a file and call the callback with the
        filedescriptor (NOT a perl filehandle, sorry for that, but watch
        out, this might change in the future).

        The pathname passed to "aio_open" must be absolute. See API NOTES,
        above, for an explanation.

        The $mode argument is a bitmask. See the "Fcntl" module for a list.
        They are the same as used in "sysopen".

        Example:

           aio_open "/etc/passwd", O_RDONLY, 0, sub {
              if ($_[0] >= 0) {
                 open my $fh, "<&=$_[0]";
                 print "open successful, fh is $fh\n";
                 ...
              } else {
                 die "open failed: $!\n";
              }
           };

    aio_close $fh, $callback
        Asynchronously close a file and call the callback with the result
        code.

    aio_read $fh,$offset,$length, $data,$dataoffset,$callback
    aio_write $fh,$offset,$length, $data,$dataoffset,$callback
        Reads or writes "length" bytes from the specified "fh" and "offset"
        into the scalar given by "data" and offset "dataoffset" and calls
        the callback without the actual number of bytes read (or -1 on
        error, just like the syscall).

        Example: Read 15 bytes at offset 7 into scalar $buffer, strating at
        offset 0 within the scalar:

           aio_read $fh, 7, 15, $buffer, 0, sub {
              $_[0] >= 0 or die "read error: $!";
              print "read <$buffer>\n";
           };

    aio_readahead $fh,$offset,$length, $callback
        Asynchronously reads the specified byte range into the page cache,
        using the "readahead" syscall.

        readahead() populates the page cache with data from a file so that
        subsequent reads from that file will not block on disk I/O. The
        $offset argument specifies the starting point from which data is to
        be read and $length specifies the number of bytes to be read. I/O is
        performed in whole pages, so that offset is effectively rounded down
        to a page boundary and bytes are read up to the next page boundary
        greater than or equal to (off-set+length). aio_readahead() does not
        read beyond the end of the file. The current file offset of the file
        is left unchanged.

    aio_stat $fh_or_path, $callback
    aio_lstat $fh, $callback
        Works like perl's "stat" or "lstat" in void context. The callback
        will be called after the stat and the results will be available
        using "stat _" or "-s _" etc...

        The pathname passed to "aio_stat" must be absolute. See API NOTES,
        above, for an explanation.

        Currently, the stats are always 64-bit-stats, i.e. instead of
        returning an error when stat'ing a large file, the results will be
        silently truncated unless perl itself is compiled with large file
        support.

        Example: Print the length of /etc/passwd:

           aio_stat "/etc/passwd", sub {
              $_[0] and die "stat failed: $!";
              print "size is ", -s _, "\n";
           };

    aio_unlink $pathname, $callback
        Asynchronously unlink (delete) a file and call the callback with the
        result code.

    aio_fsync $fh, $callback
        Asynchronously call fsync on the given filehandle and call the
        callback with the fsync result code.

    aio_fdatasync $fh, $callback
        Asynchronously call fdatasync on the given filehandle and call the
        callback with the fdatasync result code.

BUGS
    This module has been extensively tested in a large and very busy
    webserver for many years now.

       - aio_open gives a fd, but all other functions expect a perl filehandle.

SEE ALSO
    Coro, IO::AIO.

AUTHOR
     Marc Lehmann <schmorp@schmorp.de>
     http://home.schmorp.de/



Changes
Revision history for Linux::AIO

This module has been mostly superseded by IO::AIO, which is API
compatible.

* works on: ia32, amd64, alpha, sparc64.
* likely works on: s390, m68k, arm.
* unlikely to work on: ia64, hppa, v850, sparc32, mips64, ppc64.

1.8  Wed Aug 17 18:56:14 CEST 2005
	- return immediately from poll_wait if nreqs==0.
        - use <linux/unistd.h> and try to make it work.
        - replaced tests by better ones by Brad Fitzpatrick.
        - fix a bug where the data scalar would be truncated if
          the data offset is != 0.

1.71 Sun Jul 10 19:11:36 CEST 2005
	- aio_readahead was missing in 1.7.
        - xs fixes for some perls.
        - remind people to use absolute paths.
        - add reference to IO::AIO, it's brother.

1.7 Sun Jul 10 01:40:00 CEST 2005
	- further cleanups and simplifications.
        - new functions: aio_fsync, aio_fdatasync, aio_readahead.
        - vastly improved documentation.

1.6 Fri Jul  8 01:17:04 CEST 2005
        - added changes for alpha, ia64, sparc64, hppa, v850 (thanks to Stefan Traby who
          dedusted his old machines).
        - added t/02_read.t.

1.51 Sat Jul  2 15:18:21 CEST 2005
	- last release was borked.
	- killing threads increases nreqs count infinitely.
        - fixed a bug in queueing => many outstanding
          requests lead to starvation.
        - new function Linux::AIO::poll_wait.
        - documented Linux::AIO::max_parallel.

1.41 Thu Mar  3 18:00:52 CET 2005
	- change of contact address.

1.4 Sat Jan 15 01:04:27 CET 2005
        - clarify licensing (same terms as perl itself).

1.3 Sat Aug  7 17:25:31 CEST 2004
	- ported to AMD64, possibly other 64 bit architecturs
          (inspired by Joshua Hoblitt).
        - removed support for old libcs, kernels (2.2).

1.2 Sun Jul 18 12:54:43 CEST 2004
	- do not clone signal handlers.
        - when the parent thread dies, kill the child threads.
        - added aio_unlink support (Brad Fitzpatrick <brad@danga.com>)

1.1 Fri May  7 02:57:36 CEST 2004
	- the number of outstanding requests was limited by the
          pipe buffer size. this limit has been removed.
        - improve documentation of min_parallel.
        - one thread will now be started by default.

1.01 Thu May  6 14:16:08 CEST 2004
	- better support threaded perls.

1.0  Wed May  5 15:54:23 CEST 2004
	- improved the documentation and made a 1.0 release. Whoopa!

0.2  Wed May  5 12:13:02 CEST 2004
	- updates for 2.6 headers (syscalls were renamed).

0.111 Sat May 18 23:47:12 CEST 2002
	- print a notice that this module won't run on lnux <2.4.

0.11 Fri May 17 04:33:23 CEST 2002
	- actually fix the module to work on x86 (doesn't work on
          most other architectures, though).

0.1  Tue Apr  2 04:35:28 CEST 2002
	- use pread/pwrite instead of lseek+read/write.
        - fixed syscall prototypes.
        - add stat/lstat calls.
        - fixed max_parallel deadlock problem.
        - beginning of a testsuite.
        - fixed aio_close.
        - paths are now properly saved.

0.011 Tue Dec 25 03:02:30 CET 2001
	- make it compile with 5.6.1 and maybe older versions. Boy is it
          ugly now.

0.01  Mon Oct  8 14:57:59 CEST 2001
	- fixed syntax errors so it compiles with gcc-2.95.4 as well.

0.002
	- added Changes, aio_open, aio_close.

0.001
	- original version, a dire hack.


Makefile.PL
use ExtUtils::MakeMaker;

use Config;

print <<EOF;

WARNING: This module is architecture-dependent and will not likely work
         on lesser-used architectures (see Changes). If the module (esp.
         aio_read and aio_write) doesn't work, please report this and I'll
         fix it for you. I promise ;)

EOF

$Config{osvers} >= 2.4 or print <<EOF;

WARNING: This module requires 64 bit file offset functionality only found
         in linux-2.4 and later. You can still compile this module on
         older linuxes, but it won't run.

DEVELOPERS: It is not difficult to get rid of this limitation, ask me and
            I'll tell you what you need to change (and test, so I can add it!).

EOF

my $mm = MM->new({
    dist         => {
       PREOP	=> 'pod2text AIO.pm | tee README >$(DISTVNAME)/README; chmod -R u=rwX,go=rX . ;',
       COMPRESS	=> 'gzip -9v',
       SUFFIX	=> '.gz',
	            },
    NAME         => "Linux::AIO",
    VERSION_FROM => "AIO.pm",
});

sub MY::postamble {
   <<EOF;

# set \$VERSION in all modules
setver:
	\$(PERL) -pi -e 's/^(\\s*\\\$\$VERSION\\s*=\\s*).*\$\$/\$\${1}\$(VERSION);/' *.pm

EOF
}

# now get rid of -fpic, because gcc can't use syscall'ed closures
# from within pic code on x86.

use Config;

if ($Config{cppsymbols} =~ /__i386=1/) {
   for ($mm->{CCDLFLAGS}, $mm->{CCCDLFLAGS}, @{$mm->{RESULT}}) {
      s/\B-f(pic|PIC)\b//g
   }
}

$mm->flush;



AIO.pm
=head1 NAME

Linux::AIO - linux-specific aio implemented using clone

=head1 SYNOPSIS

 use Linux::AIO;

 # This module has been mostly superseded by IO::AIO.

=head1 DESCRIPTION

I<This module has been mostly superseded by IO::AIO, which is API 
compatible.>

This module implements asynchronous I/O using the means available to Linux
- clone. It does not hook into the POSIX aio_* functions because Linux
does not yet support these in the kernel (even as of 2.6.12, only O_DIRECT
files are supported) and even if, it would only allow aio_read and write,
not open, stat and so on.

Instead, in this module a number of (non-posix) threads are started that
execute your read/writes and signal their completion. You don't need
thread support in your libc or perl, and the threads created by this
module will not be visible to the pthreads library.

NOTICE: the threads created by this module will automatically be killed
when the thread calling min_parallel exits. Make sure you only ever call
min_parallel from the same thread that loaded this module.

Although the module will work with in the presence of other threads, it is
not reentrant, so use appropriate locking yourself.

=head2 API NOTES

All the C<aio_*> calls are more or less thin wrappers around the syscall
with the same name (sans C<aio_>). The arguments are similar or identical,
and they all accept an additional C<$callback> argument which must be
a code reference. This code reference will get called with the syscall
return code (e.g. most syscalls return C<-1> on error, unlike perl, which
usually delivers "false") as it's sole argument when the given syscall has
been executed asynchronously.

All functions that expect a filehandle will also accept a file descriptor.

The filenames you pass to these routines I<must> be absolute. The reason
is that at the time the request is being executed, the current working
directory could have changed. Alternatively, you can make sure that you
never change the current working directory.

=over 4

=cut

package Linux::AIO;

use base 'Exporter';

BEGIN {
   $VERSION = 1.8;

   @EXPORT = qw(aio_read aio_write aio_open aio_close aio_stat aio_lstat aio_unlink
                aio_fsync aio_fdatasync aio_readahead);
   @EXPORT_OK = qw(poll_fileno poll_cb min_parallel max_parallel nreqs);

   require XSLoader;
   XSLoader::load Linux::AIO, $VERSION;
}

=item Linux::AIO::min_parallel $nthreads

Set the minimum number of AIO threads to C<$nthreads>. The default is
C<1>, which means a single asynchronous operation can be done at one time
(the number of outstanding operations, however, is unlimited).

It is recommended to keep the number of threads low, as some linux
kernel versions will scale negatively with the number of threads (higher
parallelity => MUCH higher latency).

Under normal circumstances you don't need to call this function, as this
module automatically starts a single async thread.

=item Linux::AIO::max_parallel $nthreads

Sets the maximum number of AIO threads to C<$nthreads>. If more than
the specified number of threads are currently running, kill them. This
function blocks until the limit is reached.

This module automatically runs C<max_parallel 0> at program end, to ensure
that all threads are killed and that there are no outstanding requests.

Under normal circumstances you don't need to call this function.

=item $fileno = Linux::AIO::poll_fileno

Return the I<request result pipe filehandle>. This filehandle must be
polled for reading by some mechanism outside this module (e.g. Event
or select, see below). If the pipe becomes readable you have to call
C<poll_cb> to check the results.

See C<poll_cb> for an example.

=item Linux::AIO::poll_cb

Process all outstanding events on the result pipe. You have to call this
regularly. Returns the number of events processed. Returns immediately
when no events are outstanding.

You can use Event to multiplex, e.g.:

   Event->io (fd => Linux::AIO::poll_fileno,
              poll => 'r', async => 1,
              cb => \&Linux::AIO::poll_cb);

=item Linux::AIO::poll_wait

Wait till the result filehandle becomes ready for reading (simply does a
select on the filehandle. This is useful if you want to synchronously wait
for some requests to finish).

See C<nreqs> for an example.

=item Linux::AIO::nreqs

Returns the number of requests currently outstanding.

Example: wait till there are no outstanding requests anymore:

   Linux::AIO::poll_wait, Linux::AIO::poll_cb
      while Linux::AIO::nreqs;

=item aio_open $pathname, $flags, $mode, $callback

Asynchronously open or create a file and call the callback with the
filedescriptor (NOT a perl filehandle, sorry for that, but watch out, this
might change in the future).

The pathname passed to C<aio_open> must be absolute. See API NOTES, above,
for an explanation.

The C<$mode> argument is a bitmask. See the C<Fcntl> module for a
list. They are the same as used in C<sysopen>.

Example:

   aio_open "/etc/passwd", O_RDONLY, 0, sub {
      if ($_[0] >= 0) {
         open my $fh, "<&=$_[0]";
         print "open successful, fh is $fh\n";
         ...
      } else {
         die "open failed: $!\n";
      }
   };

=item aio_close $fh, $callback

Asynchronously close a file and call the callback with the result code.

=item aio_read  $fh,$offset,$length, $data,$dataoffset,$callback

=item aio_write $fh,$offset,$length, $data,$dataoffset,$callback

Reads or writes C<length> bytes from the specified C<fh> and C<offset>
into the scalar given by C<data> and offset C<dataoffset> and calls the
callback without the actual number of bytes read (or -1 on error, just
like the syscall).

Example: Read 15 bytes at offset 7 into scalar C<$buffer>, strating at
offset C<0> within the scalar:

   aio_read $fh, 7, 15, $buffer, 0, sub {
      $_[0] >= 0 or die "read error: $!";
      print "read <$buffer>\n";
   };

=item aio_readahead $fh,$offset,$length, $callback

Asynchronously reads the specified byte range into the page cache, using
the C<readahead> syscall.

readahead() populates the page cache with data from a file so that
subsequent reads from that file will not block on disk I/O. The C<$offset>
argument specifies the starting point from which data is to be read and
C<$length> specifies the number of bytes to be read. I/O is performed in
whole pages, so that offset is effectively rounded down to a page boundary
and bytes are read up to the next page boundary greater than or equal to
(off-set+length). aio_readahead() does not read beyond the end of the
file. The current file offset of the file is left unchanged.

=item aio_stat  $fh_or_path, $callback

=item aio_lstat $fh, $callback

Works like perl's C<stat> or C<lstat> in void context. The callback will
be called after the stat and the results will be available using C<stat _>
or C<-s _> etc...

The pathname passed to C<aio_stat> must be absolute. See API NOTES, above,
for an explanation.

Currently, the stats are always 64-bit-stats, i.e. instead of returning an
error when stat'ing a large file, the results will be silently truncated
unless perl itself is compiled with large file support.

Example: Print the length of F</etc/passwd>:

   aio_stat "/etc/passwd", sub {
      $_[0] and die "stat failed: $!";
      print "size is ", -s _, "\n";
   };

=item aio_unlink $pathname, $callback

Asynchronously unlink (delete) a file and call the callback with the
result code.

=item aio_fsync $fh, $callback

Asynchronously call fsync on the given filehandle and call the callback
with the fsync result code.

=item aio_fdatasync $fh, $callback

Asynchronously call fdatasync on the given filehandle and call the
callback with the fdatasync result code.

=cut

min_parallel 1;

END {
   max_parallel 0;
}

1;

=back

=head1 BUGS

This module has been extensively tested in a large and very busy webserver
for many years now.

   - aio_open gives a fd, but all other functions expect a perl filehandle.

=head1 SEE ALSO

L<Coro>, L<IO::AIO>.

=head1 AUTHOR

 Marc Lehmann <schmorp@schmorp.de>
 http://home.schmorp.de/

=cut

